pkg/controller/stacks/stack/stack.go

```diff
@@ -235,102 +257,237 @@ func copyLabels(labels map[string]string) map[string]string {
-		// or crossplane:ns:default:view
-		aggregationLabel := fmt.Sprintf(stacks.LabelAggregateFmt, crossplaneScope, persona)
-		labelsCopy[aggregationLabel] = "true"
+			crdPatch := client.MergeFrom(crds[i].DeepCopy())
 
-		// Each ClusterRole needs persona specific rules for each CRD
-		rules := []rbacv1.PolicyRule{}
+			labels[labelNamespace] = namespaceMember
+			crds[i].SetLabels(labels)
 
-		for _, crd := range crds {
-			kinds := []string{crd.Spec.Names.Plural}
+			h.log.Debug("adding labels for CRD", "labelNamespace", labelNamespace, "name", crds[i].GetName())
+			if err := h.kube.Patch(ctx, &crds[i], crdPatch); err != nil {
+				return err
+			}
+		}
+		return nil
+	}
+}
 
-			if subs := crd.Spec.Subresources; subs != nil {
-				if subs.Status != nil {
-					kinds = append(kinds, crd.Spec.Names.Plural+"/status")
-				}
-				if subs.Scale != nil {
-					kinds = append(kinds, crd.Spec.Names.Plural+"/scale")
-				}
+// createMultipleParentLabelsCRDHandler provides a handler which labels CRDs
+// with the stacks they are managed by. This will allow for a single Namespaced
+// stack to be installed in multiple namespaces, or different stacks (possibly
+// only differing by versions) to provide the same CRDs without the risk that a
+// single StackInstall removal will delete a CRD until there are no remaining
+// stack parent labels.
+func (h *stackHandler) createMultipleParentLabelsCRDHandler() crdHandler {
+	labelMultiParent := fmt.Sprintf(stacks.LabelMultiParentFormat, h.ext.GetNamespace(), h.ext.GetName())
+
+	return func(ctx context.Context, crds []apiextensions.CustomResourceDefinition) error {
+		for i := range crds {
+			labels := crds[i].GetLabels()
+
+			if labels[stacks.LabelKubernetesManagedBy] != "stack-manager" {
+				continue
 			}
 
-			rules = append(rules, rbacv1.PolicyRule{
-				APIGroups: []string{crd.Spec.Group},
-				Resources: kinds,
-				Verbs:     roleVerbs[persona],
-			})
-		}
+			if labels[labelMultiParent] == activeParentStack {
+				continue
+			}
 
-		// Assemble and create the ClusterRole
-		cr := &rbacv1.ClusterRole{
-			ObjectMeta: metav1.ObjectMeta{
-				Name:   name,
-				Labels: labelsCopy,
-			},
-			Rules: rules,
+			crdPatch := client.MergeFrom(crds[i].DeepCopy())
+
+			labels[labelMultiParent] = activeParentStack
+			crds[i].SetLabels(labels)
+
+			h.log.Debug("adding labels for CRD", "labelMultiParent", labelMultiParent, "name", crds[i].GetName())
+			if err := h.kube.Patch(ctx, &crds[i], crdPatch); err != nil {
+				return err
+			}
 		}
+		return nil
+	}
+}
+
+// createPersonaClusterRolesCRDHandler provides a handler which creates admin,
+// edit, and view clusterroles that are namespace+stack+version specific
+func (h *stackHandler) createPersonaClusterRolesCRDHandler() crdHandler {
+	labels := stacks.ParentLabels(h.ext)
+
+	return func(ctx context.Context, crds []apiextensions.CustomResourceDefinition) error {
+
+		for persona := range roleVerbs {
```

Comment:
this func looks like a lot of new logic from the diff, but I think that's probably not the case and it's mostly moved/refactored logic.  This section of permissions management is the one i'd be the most concerned about for regression risk, so hopefully we have done the testing validation to ensure this still works as expected.

Reply:
I agree that the refactor here is hard to follow by diff.

#### Before

`create` calls `processRBAC`
`processRBAC` was calling `createPersonaClusterRoles`
`createPersonaClusterRoles` was fetching all CRDs using `crdsFromStack` before doing role specific work
`crdsFromStack` called `crdListsDiffer` which was a naive `len(fetched) == len(stack.spec.crds)` check

#### After

`create` calls `processRBAC` (no change)
`create` calls `processCRDs` (new)
`processRBAC` no longer calls `createPersonaClusterRoles`
`processCRDs` fetches all CRDs using `crdsFromStack` (improved for multi-parented and unmanaged CRDs)
`processCRDs` walks a list of handlers which include:
* `createPersonaClusterRoles` (inner crd-loop is unmodified, outer crd fetch + loop + len-check is not needed)
* multi-parent labeller (new)
* namespace discovery label (moved here away from unpack, because the managed crd may already exist)

`crdsFromStack` now verifies that the expected version of the CRD exists the fetched CRD
`crdsFromStack` is now usable for deletion where some CRDs could already be missing
`processCRDs` uses `crdListFulfilled` to verify that all expected CRDs are present (managed or not)